Code/Framework/AzCore/AzCore/IO/ByteContainerStream.h

```diff
@@ -76,6 +86,9 @@ namespace AZ
             ContainerType*  m_buffer;
             size_t          m_pos;
             size_t          m_maxGrowSize; //< Maximum grow size to avoid the buffer growing too fast when it's working on big files
+            //! Used to track if the ByteContainerStream has received a Close() call
+            //! By default the ByteContainerStream is open
+            bool            m_opened{ true };
```

Comment:
This is necessary because the checking if the stream is open via the `AZ::IO::GenericStream::IsOpen` function would return true even if a `Close()` call was performed.

This was causing an issue in the `ArchiveWrite::UnmountArchive` as it invokes Commit to write the Archive Header and Table of Contents if the file was open, which it would continuously do even if a previous unmount call has occurred.

>1. It should probably get set to true by the constructor that takes in a buffer pointer instead of defaulting to open, so that way if other constructors (like default constructors) get added, they would correctly start as closed, same as the other types of streams.

Well I defaulted it to true in the member initializer due to the maintaining the current the ByteContainerStream API is working with being associated with a single stream from the constructor.
I do like the flexibility however supporting an `Open` function that would take in reference to the buffer type, in which case the state a container being "open" is no longer tied to construction.
With that change it would make sense to no longer default the `m_opened` to true and instead set it in the constructor/`Open`
 

>2. All of the operations (seek / read / write) should handle the case that they're called when the stream isn't open. Either assert, return back an error or noop, or both.

This was kept inline with the other GenericStream implementations where their Seek, Read and Writes don't check if a file is open before performing the operations: https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/GenericStreams.cpp#L223-L238

Reply:
Yeah, it's not done in GenericStream, but there's some amount of checking in general:
https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/SystemFile.cpp#L118
https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/GenericStreams.cpp#L278-L281

I guess what I was thinking is that Close() should probably clear out the m_buffer pointer, and read/write/seek should be accounting for this, because otherwise you can keep reading/writing even though it's "closed" making the tracking of "closed" not really correct or functional.

Reply:
I'll update the ByteContainerStream, to have proper lifetime semantics, so that when the stream is not in the "open state", it will no-op if trying to seek, read or write.